Spring Web으로 todo Rest Api 만들기

kckoh opened the pull request:
리뷰 부탁드립니다.
구현:
    CRUD API
    PASSED API Testing
고려해야할 점:
    커스텀 Exception 클라스를 만들었는데, 과연 지금 필요한지 모르겠네요.

kckoh committed:
Pass API testing

리뷰 부탁드립니다.
구현:
    CRUD API
    PASSED API Testing
고려해야할 점:
    커스텀 Exception 클라스를 만들었는데, 과연 지금 필요한지 모르겠네요.

johngrib commented on app/src/main/java/com/codesoom/assignment/type/ErrorCode.java:
```diff
@@ -0,0 +1,11 @@
+import lombok.Getter;
+
+@Getter
+@AllArgsConstructor
```
롬복의 `@AllArgsConstructor`는 널리 알려진 위험성이 있습니다. 가급적이면 이 애노테이션 없이 순수한 자바 코드로 해결해 보도록 합시다. 어떤 팀이나 회사에서는 이 애노테이션의 사용을 금지하는 경우도 많아요. 어째서 그러한지에 대해서도 조사해 보세요.

kckoh replied:
ㅅ찾아 보니까 IDE가 제공해주는 리팩토링이  안되고, compile 할때 생성자의 파라미터 순서를 필드 선언 순서에 맞춰 변경해버린다고 나와 있네요! 그래서 `@AllArgsConstructor는` 안쓰기로 했습니다!

johngrib replied:
스스로 안 쓰기로 결정하셔서 기쁩니다. 제가 쓰지 마세요! 하는 것보다 직접 결정하신 것이 더 의미가 있다고 생각해요. 한편 `@`를 그대로 사용하면 다른 github 사용자가 연결되어 해당 사용자에게 알림이 가게 됩니다.

![image](https://user-images.githubusercontent.com/1855714/191270365-c5ea7c6a-6c43-452a-b529-b7ac8f0fece9.png)

그래서 이렇게 직접 `@`를 쓰는 것보다 양쪽에 백틱을 넣어서 그 사용자를 링크하지 않게 해주세요.

```
`@AllArgsConstructor`
```

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+
+    private Long generateId( ) {
+        newId += 1;
+        return newId;
+    }
```
여기에 `synchronized` 키워드를 적용할 필요가 있을지에 대해 고민해 보세요.

kckoh replied:
여기에는 synchronized 키워드를 적용하였습니다!

kckoh replied:
synchronized 키워드로 적용 하였습니다!

johngrib replied:
하는 김에 JLS의 `synchronized` 섹션도 꼭 읽어보세요. 항상 공식 문서를 읽는 습관을 길러두시고요.

https://docs.oracle.com/javase/specs/jls/se17/html/jls-14.html#jls-14.19

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,79 @@
+
+    @GetMapping("/{id}")
+    public ResponseEntity<Task> getTaskId(@PathVariable Long id) throws TaskException {
+        for (Task task : tasks
+        ) {
+            if (task.getId() == id) {
+                return new ResponseEntity<>(task, HttpStatus.OK);
+
+            }
+        }
```
만약 tasks의 길이가 320만 정도 된다면 어떤 문제가 있을까요? 서비스를 오래 운영하게 된다면 아마 tasks의 앞쪽은 오래된 task들로 차있을 것이고, 오래된 것들은 거의 조회하지 않게 될 겁니다. 빈번하게 조회되는 것들은 최근에 추가된 것들이겠죠.

kckoh replied:
concurrentHashMap으로 o(n)에서 o(1)으로 바꿧습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,79 @@
+@RequestMapping("/tasks")
+@CrossOrigin
+public class TaskController {
+    private List<Task> tasks = new ArrayList<>();
```
`new ArrayList`를 갓 만들었을 때 `ArrayList`의 capacity는 얼마인까요? 만약 잘 모르겠다면 이번 기회에 `ArrayList` 클래스를 열어서 조사해 봅시다.

kckoh replied:
보니까 initial capacity가 10으로 되어 있네요! 그래서 concurrentHashMap를 써서 필요할때 arrayList로 변경 하기로 했는데 이렇게 구현하면 문제점이 있을까요?

johngrib replied:
특별한 문제가 있지는 않아요. 그러나 ArrayList의 initial capacity와 증감에 대해 크게 신경쓰지 않고 코딩하는 경우가 많습니다. 이번 기회에 Java 애플리케이션 개발자들이 매일 사용하는 ArrayList에 대해 확실히 공부해두면 좋을 거라 생각해서 언급해 드렸어요. 시간날 때 Dynamic Array 자료구조에 대해 조사해보시고, Java의 ArrayList에서는 어떻게 구현했는지에 대해서도 연구해 보세요. Java의 ArrayList는 꽤나 훌륭한 결정이 많이 들어간 클래스이기 때문에 코드를 읽으면 읽을수록 배우실 수 있는 게 많을 거에요.

kckoh replied:
간략하게 dynamic array 에 대해서 적었습니다! https://kckoh.hashnode.dev/java-dynamic-arrays

johngrib replied:
좋습니다! grow 까지 잘 살펴보셨네요!

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,79 @@
+    @ResponseStatus(HttpStatus.CREATED)
+    public Task create(@RequestBody Task task) {
+
+        task.setId(generateId());
```
동시에 `POST` 요청이 3000개 정도 들어오면 `generateId` 메소드는 제대로 작동하게 될까요? 즉 모두 다른 3000개의 아이디를 생성하는데 성공할 수 있을까요?

kckoh replied:
create()메소드와 generateId() 메소드에 synchronized keyword를 적용 하였습니다. 이렇게 하면 문제점이 있을까요?

johngrib replied:
지금은 이 정도로 충분할 것 같습니다.

johngrib reviewed (commented):
반갑습니다 :-)

kckoh committed:
fix: refactor some code and pass both e2e and api test

고친점:
    @AllArgsConstructor 지우기
    generateId() method synchronized keyword 추가
    ForEach -> hashMap get funciton
    refactor arrayList to concurrentHashMap
    create() method synchronized keyword 추가

johngrib commented on app/src/main/java/com/codesoom/assignment/configuraiton/WebConfig.java:
```diff
@@ -0,0 +1,18 @@
+                .allowedMethods("*");
+//                .allowedMethods(HttpMethod.GET.name());
+    }
+}
```
파일 마지막에 공백 라인 하나를 추가해 주세요. 파일이 `}` 와 같이 공백이 아닌 문자로 끝나게 되면 github에서 다음과 같이 경고가 나옵니다.

![image](https://user-images.githubusercontent.com/1855714/191270555-ccef17df-b465-4a27-91cb-279802484b4d.png)

이것이 문제가 되는 이유는 POSIX 표준에서 Line을 "개행 문자로 끝나는 0개 이상의 개행 문자가 아닌 문자들의 시퀀스"으로 정의하고 있기 때문입니다.

만약 파일을 작성할 때마다 마지막에 공백 라인을 추가하는 것이 귀찮다면, IntelliJ에 관련 설정을 해주는 방법이 있습니다.

다음은 제 블로그에 설정 방법을 기록한 것입니다. 참고해주세요.
https://johngrib.github.io/wiki/intellij/#%ED%8C%8C%EC%9D%BC-%EB%A7%88%EC%A7%80%EB%A7%89%EC%97%90-%EC%9E%90%EB%8F%99%EC%9C%BC%EB%A1%9C-1%EA%B0%9C%EC%9D%98-%EA%B0%9C%ED%96%89-%EB%AC%B8%EC%9E%90-%EC%B6%94%EA%B0%80%ED%95%98%EA%B8%B0

kckoh replied:
추가 하였습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -1,79 +1,91 @@
+        return new ResponseEntity<>(task, HttpStatus.OK);
+
 
+//        for (Task task : tasks) {
+//            if (task.getId() == id) {
+//                return new ResponseEntity<>(task, HttpStatus.OK);
+//            }
+//        }
+//        return new ResponseEntity<>(new Task(), HttpStatus.NOT_FOUND);
```
이런 코드 조각에 대해 주석 처리를 하는 것보다 그냥 삭제하는 습관을 들여보세요. 삭제하기 전에 commit이 남겨져 있으니 코드가 사라지지 않는다는 점을 염두에 두세요.

`git show` 명령에 대해 조사해 보시고, 익숙해질 때까지 터미널에서 직접 반복해서 사용해 보세요.

kckoh committed:
refactor TaskController

고친점:
    delete commented code
    add extra space for all files

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,72 @@
+import com.codesoom.assignment.models.Task;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
```
코드숨 과정을 진행하는 동안에는 `import *`의 사용을 지양하도록 합시다. 어떤 것을 import하는지 분명히 명시하고 눈으로 익혀두는 연습을 해보자는 의미이기도 해요.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+
+    @PostMapping
```
추가 과제입니다. RFC 7231 문서에서 Post method가 있는 부분을 찾아 읽어보시고 글로 정리해 보세요. 물론 여유가 있으시다면 다른 메소드도 같이 읽고 정리해보시면 더 좋겠죠!

kckoh replied:
간략하게 블로그에 정리 하였습니다.!
https://kckoh.hashnode.dev/rfc-7231

johngrib replied:
좋습니다. GET, POST 외의 다른 메소드들도 시간나실 때 학습해 보세요. 모두 한번씩은 RFC 문서를 훑어보셔야 합니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.web.bind.annotation.CrossOrigin;
+
+@Getter
+@Setter
```
추가 과제입니다. `@Setter`와 `setId`, `setTitle` 메소드 없이 `Task` 클래스 코드를 다시 디자인해보세요. 추가 과제이므로 안 하셔도 괜찮습니다. 하지만 해보시길 권합니다 :-)

kckoh committed:
refactor: apply builder pattern to remove setter

고친점:
    apply Builder Pattern in Task.java
    blog post about RFC 7231 Post request

johngrib commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -1,14 +1,47 @@
     private Long id;
     private String title;
+
+    public Task() {
+    }
+
+    public Task(Long id, String title) {
+        this.id = id;
+        this.title = title;
+    }
+
+    public static TaskBuilder builder() {
+        return new TaskBuilder();
+    }
+
+    public static class TaskBuilder {
+        private Long id;
+        private String title;
+
+        TaskBuilder() {
+        }
+
+        public TaskBuilder id(Long id) {
+            this.id = id;
+            return this;
+        }
+
+        public TaskBuilder title(String title) {
+            this.title = title;
+            return this;
+        }
+
+        public Task build() {
+            return new Task(this.id, this.title);
+        }
+    }
```
이제 setter가 없는데 title을 수정해야 할 일이 있다면 어떻게 해야 할까요?

kckoh replied:
updateTitle(String title) 메소드를 만들었습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,71 @@
+import com.codesoom.assignment.models.Task;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
```
코드숨 과정을 진행하는 동안에는 `import *`의 사용을 지양하도록 합시다.

kckoh replied:
혹시 `import  *` 지양하는 이유를 알 수 있을까요?

johngrib replied:
`import *`를 사용하고 안하고는 회사마다 팀마다 컨벤션에 의해 달라질 수 있어요. 어느쪽이 더 좋거나 나쁘거나 하지는 않습니다. 코드숨에서는 `import *`를 사용하지 않고 어떤 클래스를 import하는지 명시하는 것을 목표로 삼는 컨벤션이 있다고 생각해주시면 되겠습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,71 @@
+        return new ResponseEntity<>(new Task(), HttpStatus.NOT_FOUND);
+    }
+
+    private synchronized Long generateId() {
+        newId += 1;
+        return newId;
+    }
```
`IdGenerator` 라는 이름을 가진 클래스를 만들고 아이디 생성 로직을 분리해 보세요. (더 나은 이름이 있다면 그 이름으로 하셔도 됩니다)

kckoh replied:
분리 하였습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/exception/TaskException.java:
```diff
@@ -0,0 +1,12 @@
+
+@Getter
+@Setter
+public class TaskException extends Throwable {
```
Throwable은 범위가 너무 넓습니다. `java.lang.Error`도 Throwable에 포함되죠. 이번 기회에 자바의 Throwable 상속 구조에 대해서도 조사해 보세요.

```suggestion
public class TaskException extends RuntimeException {
```

kckoh replied:
블로그에 남겼습니다. https://kckoh.hashnode.dev/throwable-class-and-hierarchy

kckoh committed:
refactor: IdGenerator and create repository and service

고친 점:
    Repository & service 분리
    IdGenerator 클라스 분리
    import * 지양하기
    Throwable 바꾸기

johngrib reviewed (commented):
작업을 잘 하셨습니다. 그런데 하나의 커밋에 너무 여러 주제가 합쳐져 있어요.

![image](https://user-images.githubusercontent.com/1855714/191965387-206f8c3d-97d5-461a-81f1-9c4e255ced53.png)

이 목록에 있는 한 줄 한 줄이 각각 하나의 커밋이 되었다면 더 좋았을 겁니다.
다음부터는 하나의 커밋이 하나의 주제를 표현하도록 작업해보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -0,0 +1,54 @@
+        this.title = title;
+    }
+
+    public void updateTitle(String title) {
+        this.title = title;
+    }
```
이건 이름에만 `set`이 아닌 setter입니다. 우리가 무엇을 목표로 setter 를 사용하지 않기로 하게 된 것인지를 생각해 보셔야 해요.
목표가 불변 객체이므로 다음과 같이 작업하는 것을 고려해보세요.

```suggestion
    public Task updateTitle(String title) {
        return new Title(this.id title);
    }
```

물론 메소드 이름은 더 나은 이름이 가능합니다. 이름에 대해서도 고민해 보세요.
그리고 이렇게 하게 되면 어떤 장단점이 있는지에 대해서도 생각해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/repository/IdGenerator.java:
```diff
@@ -0,0 +1,10 @@
+package com.codesoom.assignment.repository;
+
+public class IdGenerator {
+    private static Long newId = 0L;
+
+    public static synchronized Long generateId() {
+        newId += 1;
+        return newId;
+    }
+}
```
좋습니다. 이제 JavaDoc 작성방법을 조사해보시고, 이 클래스에 JavaDoc 주석을 작성해 보세요.

kckoh replied:
간략하게 블로그에 적었습니다. johngrib님의 블로그 참고 하였습니다!
https://kckoh.hashnode.dev/javadoc

kckoh committed:
add: make Task class as immutable object

고친점:
    Task 클라스를 불변객체로 만들었습니다.
    final 키워드를 이용해서 구현 하였습니다.

kckoh committed:
refactor: add a javadocs to IdGenerator

고친점:
    간단하게 javadocs를 사용하여서 api documentation를 만들었습니다.

kckoh committed:
fix: remove a parameter from generatedId()

고친점:
    CI 테스트를 통과 하지 못해서 고쳤습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -9,19 +9,21 @@
+        this.title = "";
+    }
+
+    public Task updateTitle(String title) {
+        return new Task(this.id, title);
```
잘 하셨습니다. 다만 이런 기법이 있다는 정도로 알아두시고, 호불호가 갈리기 때문에 흔하게 사용되는 패턴이 아니라는 것도 기억해 두세요.

johngrib reviewed (commented):
4ae6562761a4da9484a8ec16cb4047faba18f2b0 커밋에 빌드된 결과물이 포함되어 있습니다. 빌드 결과는 포함시키지 않도록 작업해 주시거나 git을 사용해서 커밋을 분리하는 방법을 조사해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/repository/IdGenerator.java:
```diff
@@ -0,0 +1,19 @@
+public class IdGenerator {
+    private static Long newId = 0L;
+
+    /**
+     * Returns a newly generated id
+     *
+     * @return newId
+     */
+
```
```suggestion
    /**
     * Returns a newly generated id.
     *
     * @return newId
     */
```

johngrib commented on app/src/main/java/com/codesoom/assignment/repository/TaskService.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.List;
+import java.util.Optional;
+
+@Service
+public class TaskService {
```
좋습니다. 이제 controller, service, repository 처럼 중요한 컴포넌트의 모든 public 메소드에 JavaDoc을 작성해 보세요. 작성해 보시다 보면 뭔가 이상한 것, 개선하고 싶은 것들이 보이게 될 수도 있습니다. 그런 것이 있다면 어떤 것들이었는지에 대해서도 이야기해 보도록 합시다.

kckoh replied:
javadocs를 작성하면서 개선하고 싶은것이 있어서 리펙토링을 하였습니다.ResponseEntity를 TaskService와 TaskRepository 메소드에서 리턴을 하고 있어서 모든 ResponseEntity를 TaskService 메소드에서 리턴하도록 리펙토링 하였습니다.

johngrib replied:
좋습니다! ResponseEntity는 HTTP 프로토콜과 관련된 것이기 때문에 가급적이면 컨트롤러 레이어를 벗어나지 않는 쪽으로 작업하는 것이 좋죠.

johngrib commented on javaDocs/allclasses-frame.html:
```diff
@@ -0,0 +1,29 @@
+<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
+<!-- NewPage -->
+<html lang="en">
+<head>
+<!-- Generated by javadoc (1.8.0_281) on Fri Sep 23 22:28:36 PDT 2022 -->
+<title>All Classes</title>
+<meta name="date" content="2022-09-23">
+<link rel="stylesheet" type="text/css" href="stylesheet.css" title="Style">
+<script type="text/javascript" src="script.js"></script>
+</head>
+<body>
+<h1 class="bar">All&nbsp;Classes</h1>
+<div class="indexContainer">
+<ul>
+<li><a href="com/codesoom/assignment/App.html" title="class in com.codesoom.assignment" target="classFrame">App</a></li>
+<li><a href="com/codesoom/assignment/type/ErrorCode.html" title="enum in com.codesoom.assignment.type" target="classFrame">ErrorCode</a></li>
+<li><a href="com/codesoom/assignment/controllers/HelloController.html" title="class in com.codesoom.assignment.controllers" target="classFrame">HelloController</a></li>
+<li><a href="com/codesoom/assignment/repository/IdGenerator.html" title="class in com.codesoom.assignment.repository" target="classFrame">IdGenerator</a></li>
+<li><a href="com/codesoom/assignment/models/Task.html" title="class in com.codesoom.assignment.models" target="classFrame">Task</a></li>
+<li><a href="com/codesoom/assignment/models/Task.TaskBuilder.html" title="class in com.codesoom.assignment.models" target="classFrame">Task.TaskBuilder</a></li>
+<li><a href="com/codesoom/assignment/controllers/TaskController.html" title="class in com.codesoom.assignment.controllers" target="classFrame">TaskController</a></li>
+<li><a href="com/codesoom/assignment/exception/TaskException.html" title="class in com.codesoom.assignment.exception" target="classFrame">TaskException</a></li>
+<li><a href="com/codesoom/assignment/repository/TaskRepository.html" title="class in com.codesoom.assignment.repository" target="classFrame">TaskRepository</a></li>
+<li><a href="com/codesoom/assignment/repository/TaskService.html" title="class in com.codesoom.assignment.repository" target="classFrame">TaskService</a></li>
+<li><a href="com/codesoom/assignment/configuraiton/WebConfig.html" title="class in com.codesoom.assignment.configuraiton" target="classFrame">WebConfig</a></li>
+</ul>
+</div>
+</body>
+</html>
```
javaDocs/ 의 모든 파일은 커밋에서 제외시켜 주세요.

kckoh replied:
gitignore에서 javdocs를 추가 했습니다.

kckoh committed:
add: gitignore for javadocs

고친점:
    javadocs가 빌드 안되게 gitignore 추가

kckoh committed:
add: javadocs for TaskController, TaskService and TaskRepository

고친점:
     javadocs를 추가 하였습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,86 @@
+
+import java.util.List;
+
+/**
+ * The class is responsible for mapping /tasks requests
+ */
```
`The class is..` 로 설명하지 않고 좀 더 단순하게 설명해봅시다.

kckoh replied:
https://kckoh.hashnode.dev/september-25-2022-retrospectives 회고 적었습니다. 다음주도 열심히 해보겠습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,86 @@
+        this.taskService = taskService;
+    }
+
+    /**
+     * Returns all lists of tasks from the database
+     * @return the List of tasks
+     */
```
좋습니다.
```suggestion
    /**
     * Returns all lists of tasks from the database.
     * @return the List of tasks
     */
```

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,86 @@
+     *
+     * Returns ResponseEntity with the found task
+     * @param id from URI parameter
+     * @return ResponseEntity with HttpStatus.OK; otherwise, returns ResponseEntity with HttpStatus.NOTFOUND
```
간단하게 표현해보면 어떨까요? `@return`의 내용은 핵심적인 내용을 짧게 표현해 보세요.

```suggestion
     * @return 조회 결과
```

johngrib commented on app/src/main/java/com/codesoom/assignment/repository/IdGenerator.java:
```diff
@@ -0,0 +1,18 @@
+public class IdGenerator {
+    private static Long newId = 0L;
+
+    /**
+     * Returns a newly generated id.
+     *
+     * @return newId
+     */
```
좋습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/repository/IdGenerator.java:
```diff
@@ -0,0 +1,18 @@
+package com.codesoom.assignment.repository;
+
+/**
+ * This class stores id for Task and manipuldates id
+ */
```
id 를 "저장하고 있다"는 것은 굳이 설명하지 않아도 됩니다. 만약 이 클래스의 구현을 변경해서 id 를 저장하지 않고 랜덤하게 생성하도록 고쳤다고 가정해 봅시다. 그렇게 되면 이 주석도 같이 고쳐줘야 하겠죠? "아이디를 생성해준다"는 이 클래스의 책임만 단순하고 명확하게 설명해주시면 됩니다.

johngrib reviewed (commented):
빠르게 학습해나가시는 것이 인상적인 한 주였어요. 다음주도 화이팅!
주간 회고를 작성해 공유해 주시면 머지해드리겠습니다.

kckoh committed:
fix: clarifying javadocs documentation

고친점:
    javadocs를 좀더 간결하게 적었습니다.
